convox 3.0.14

ddollar opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

SMillerDev commented on Formula/convox.rb:
```diff
@@ -1,8 +1,8 @@
   homepage "https://convox.com/"
-  url "https://github.com/convox/rack/archive/20200129210159.tar.gz"
-  sha256 "945d6237ea0108ef0cd5ab9e6dae0f898710ca84a65d8480b8466e4c40ff99af"
+  url "https://github.com/convox/convox/archive/3.0.0.tar.gz"
```
Since this is a very different versioning scheme you should add a line `version_scheme 1`

SMillerDev commented on Formula/convox@2.rb:
```diff
@@ -0,0 +1,39 @@
+class Convox < Formula
```
This won't work. In addition to this, will v2 be supported by upstream for a while next to v3?

SMillerDev commented on Formula/convox@2.rb:
```diff
@@ -0,0 +1,39 @@
+  url "https://github.com/convox/rack/archive/20200129210159.tar.gz"
+  sha256 "945d6237ea0108ef0cd5ab9e6dae0f898710ca84a65d8480b8466e4c40ff99af"
+
+  bottle do
```
Don't add bottles to new formulae.

ddollar commented:
Thanks @SMillerDev ! I have updated the PR with your suggested changes.

ddollar commented:
Thanks for the review @zbeekman 

Unfortunately all of the commands (including the ones you mentioned) require authentication with a remove API. I have modified the test to test the version output of `convox -v` is as expected which should validate that the CLI is working as much as can be without access to a remote API.

zbeekman commented:
> Unfortunately all of the commands (including the ones you mentioned) require authentication with a remote API.

Could you test that the attempted authentication fails? You could even point it to a non-existent endpoint. That way you can see it is attempting to connect/authenticate and assert that it will fail. Would that be possible at all?

zbeekman commented:
BTW, your current test is better than it was, but it would be nice if you could rebase and fixup---i.e., squash into a single commit---so if you could check that it attempts to authenticate and fails (with the expected output) that would be an improvement.

ddollar commented:
Rebased and squashed. I gave some thought to going the auth/failed route but it seems like it would make the test flaky and dependent on local machine state (is the user already logged in?)

zbeekman commented:
> I gave some thought to going the auth/failed route but it seems like it would make the test flaky and dependent on local machine state (is the user already logged in?)

Yes, as far as I know the user is already logged in. When you open up the CI VMs you can see the desktop etc. But you do need to take care that you have a mechanism to prevent timeouts and cleanup after the test, especially if you spawn an asynchronous process. There are some formulae that do this, but I'm trying to recall which ones and what to search on, but I'm drawing a blank. I know formulae that start services and servers, like `oauth2_proxy` will start a server in a forked process, sleep for a bit, then query them. And then others like `ddclient` show  examples of forking a process and cleaning up for client software connecting with remote servers.

Basically, run something like:
```
find "$(brew --repo homebrew/core)/Formula" -type f -name '*ctl*'
# or
find "$(brew --repo homebrew/core)/Formula" -type f -name '*client*'
```
To find examples of how other people do this. E.g., `ddclient`, `glooctl`, `fluxctl` etc.

Some folks are using simple/bad version checks, but whenever possible we try to get them to update. If you try something and CI barfs you can always revert it if you can't figure out how to make it work.

SMillerDev commented on Formula/convox.rb:
```diff
@@ -19,21 +20,22 @@ class Convox < Formula
 
   test do
-    system bin/"convox"
+    assert_equal version, shell_output("#{bin}/convox -v").chomp
```
We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

ddollar commented:
Updated test to attempt invalid login

SMillerDev commented on Formula/convox.rb:
```diff
@@ -19,21 +20,27 @@ class Convox < Formula
 
   test do
-    system bin/"convox"
+    require "open3"
```
This is not needed, there's a homebrew dsl to read the shell output. `shell_output "<cmd>", <exitcode>`

ddollar replied:
shell_output did not appear to capture stderr

ddollar replied:
I was going off of the information found here: https://www.rubydoc.info/github/Homebrew/brew/Formula.test

SMillerDev replied:
> shell_output did not appear to capture stderr

Not by default but you can just use `cmd 2>&1` to do that.

ddollar replied:
The test seems more accurate as is

SMillerDev replied:
The test also depends on gems that not everyone might have installed and is out of line with homebrew standards like this. Just use the homebrew DSL, it's there for a reason.

ddollar replied:
Open3 is part of the Ruby standard library

https://www.rubydoc.info/stdlib/open3/Open3.popen3

SMillerDev replied:
That doesn't change the homebrew standards. It's not really up for debate, it's not getting merged like this.

zbeekman commented:
@ddollar The changes are looking good. I need to research what the official stance on `popen3` is vs capturing standard error and standard out with homebrew DSL. (Sean knows better than I do on this front so he's probably correct...) Also, if the build isn't queued by morning I'll check in on the CI system and if needed re-queue it. Feel free to ping me tomorrow if there's still no link to a Jenkins build.

SMillerDev commented:
Didn't notice you pushed the changes, could you reword the commit to `<name> <version>`? Then it'll be good to go.

SMillerDev commented:
And please tag me after, force-push is silent

ddollar commented:
@SMillerDev updaetd

SMillerDev commented:
Thanks @ddollar! Without contributions like yours it'd be impossible to keep homebrew going with the high standards that users have come to expect from the project. You can feel good knowing that you've made the world a tiny bit better for homebrew users around the world! 👍 🎉
